rfd/0135-kube-apps-discovery.md

```diff
@@ -0,0 +1,227 @@
+
+If Kubernetes service has multiple ports in use, we will expose each port as a separate app, with name of the app including the port's name
+specified in the service. For example, if a service has port 443 exposed with name `tls` and also port `5432` with name `postgres`
+we will create two apps: `testApp-tls-main-cluster-prod`, `testApp-postgres-main-cluster-prod`. Names for ports are mandatory in Kubernetes if more
+than one port is exposed for the service, so we always will be able to use it if we need it. User will be able to use annotation to specify
+preferred port, in that case only app with this port will be created, without port info in the app name.
+
+In order for discovery service to be able to dynamically create/update/delete apps we will add Read/Write permissions 
+for the apps to the built-in system role `Discovery`.
+
+Labels from the Kubernetes service will be copied to the corresponding Teleport app.
+
+## UX
+
+### Discovery
+
+Discovery service will use new `kubernetes` matchers to periodically poll list of services inside the Kubernetes cluster.
+Labels specified in the config will be used to filter out services that should be exposed. Also users will be able to
+specify namespaces to process. Another new field for the configuration of the discovery service will be `kubernetes_cluster` - 
+it specifies kubernetes cluster name in which discovery service is running. It is mandatory if `kubernetes` matchers are
+specified in the config.
+
+```yaml
+## This section configures the Discovery Service
+discovery_service:
+  enabled: yes
+  discovery_group: kube-auto-apps
+  kubernetes_cluster: main-cluster
+  kubernetes:
+    - types: ["app"] # in the future "db" will be possible
+      namespaces: [ "toronto", "porto" ]
+      labels: # List of labels to select desired Kubernetes services
+        env: staging
+    - types: ["app"]
+      namespaces: [ "seattle", "oakland" ]
+      labels:
+        env: testing
+```
+
+`kubernetes_cluster` field in the config will be translated into `teleport.dev/kubernetes-cluster` label on the Teleport app resource, so
+later it's be easier to target it in the corresponding app service.
+
+#### Annotations
+
+Kubernetes annotations will allow users to fine tune transformation of services into Teleport apps. They will override default behaviour, 
+but they are not required for import of services - by default services without any annotations will also be imported.
+
+Annotation `teleport.dev/discovery-type` controls what Teleport entity will be created from this service, for now we will only support value 
+`app` which means Teleport application will be imported from this service. In the future there are plans to expand to database importing.
+If annotation is not set service is treated as an app.
+
+Annotation `teleport.dev/discovery-protocol` controls protocol for the access of the Teleport app we create. If annotation is missing
```

Comment:
Nit: I'd shorten name a bit.
```suggestion
Annotation `teleport.dev/protocol` controls protocol for the access of the Teleport app we create. If annotation is missing
```

Reply:
Should we consider adding a namespace e.g. `discovery.teleport.dev/port`?

Reply:
Namespace sounds good to me as well, but it doesn't make shorter